engine/src/btc/cached_rpc.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{
+	btc::{
+		retry_rpc::BtcRetryRpcClient,
+		rpc::{BlockHeader, BtcRpcApi, VerboseBlock},
+	},
+	caching_client::CachingClient,
+};
+use bitcoin::{BlockHash, Txid};
+use cf_chains::btc::{BlockNumber, BtcAmount};
+use cf_utilities::{loop_select, task_scope::Scope};
+use futures::FutureExt;
+use tokio::sync::mpsc;
+
+#[derive(Clone)]
+pub struct BtcCachingClient {
+	block: CachingClient<BlockHash, VerboseBlock, BtcRetryRpcClient>,
+	block_hash: CachingClient<BlockNumber, BlockHash, BtcRetryRpcClient>,
+	send_raw_transaction: CachingClient<Vec<u8>, Txid, BtcRetryRpcClient>,
+	next_block_fee: CachingClient<(), Option<BtcAmount>, BtcRetryRpcClient>,
+	avg_fee_rate: CachingClient<BlockHash, BtcAmount, BtcRetryRpcClient>,
+	block_header: CachingClient<BlockHash, BlockHeader, BtcRetryRpcClient>,
+	best_block_hash: CachingClient<(), BlockHash, BtcRetryRpcClient>,
+
+	pub cache_invalidation_sender: mpsc::Sender<()>,
+}
+
+impl BtcCachingClient {
```

Comment:
I have a feeling that we could/should deduplicate this code. Maybe https://crates.io/crates/duplicate would be useful? Up to discussion.